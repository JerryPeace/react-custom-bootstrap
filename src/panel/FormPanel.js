import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import styleMaps from '../styleMaps';
import bootstrapUtils from '../utils/bootstrapUtils';

class FormPanel extends React.Component {
    static propTypes = {
        itemsConfig: PropTypes.object
    };

    static defaultProps = {
        bsClass: 'form',
        bsStyle: 'horizontal',
        itemsConfig: {
            standalone: false
        }
    };

    render() {
        const {
            children,
            itemsConfig,
            ...props
        } = this.props;
        const bsCls = bootstrapUtils.getCls(props);
        let newChildren;

        if (!itemsConfig) {
            newChildren = children;
        } else {
            newChildren = React.Children.map(children, function (child) {
                if (React.isValidElement(child) && typeof child.type === 'function') {
                    return React.cloneElement(child, itemsConfig);
                }
                return child;
            });
        }

        return (
            <form className={classNames(this.props.className, bsCls)}>
                {newChildren}
            </form>
        );
    }
}

export default FormPanel;
