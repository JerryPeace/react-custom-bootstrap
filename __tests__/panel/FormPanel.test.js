import React from 'react';
import { mount } from 'enzyme';
import _ from 'lodash';
import FormPanel from 'panel/FormPanel';

describe('FormPanel', () => {
    test('render FormPanel default', () => {
        const DummyComp = (props) => {
            expect(props.standalone).toBeFalsy();
            return <span>{props.children}</span>;
        };

        const form = mount(
            <FormPanel>
                <DummyComp>test</DummyComp>
            </FormPanel>
        );
        expect(form).toMatchSnapshot();
    });

    test('render FormPanel without itemsConfig', () => {
        const form = mount(
            <FormPanel itemsConfig={null}>
                <span>test</span>
            </FormPanel>
        );
        expect(form).toMatchSnapshot();
    });

    test('renderFormPanel with custom itemsConfig', () => {
        const itemsConfig = { a: 1, b: 'b', c: true };

        const DummyComp = (props) => {
            console.log('props', props);
            _.each(itemsConfig, (v, k) => {
                expect(props[k]).toBe(v);
            });
            return <span>{props.children}</span>;
        };

        const form = mount(
            <FormPanel itemsConfig={itemsConfig}>
                <DummyComp>test</DummyComp>
            </FormPanel>
        );
        expect(form).toMatchSnapshot();
    });
});
