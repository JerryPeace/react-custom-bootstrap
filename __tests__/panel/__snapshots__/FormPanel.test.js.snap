// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPanel render FormPanel default 1`] = `
<FormPanel
  bsClass="form"
  bsStyle="horizontal"
  itemsConfig={
    Object {
      "standalone": false,
    }
  }
>
  <form
    className="form form-horizontal"
  >
    <DummyComp
      standalone={false}
    >
      <span>
        test
      </span>
    </DummyComp>
  </form>
</FormPanel>
`;

exports[`FormPanel render FormPanel without itemsConfig 1`] = `
<FormPanel
  bsClass="form"
  bsStyle="horizontal"
  itemsConfig={null}
>
  <form
    className="form form-horizontal"
  >
    <span>
      test
    </span>
  </form>
</FormPanel>
`;

exports[`FormPanel renderFormPanel with custom itemsConfig 1`] = `
<FormPanel
  bsClass="form"
  bsStyle="horizontal"
  itemsConfig={
    Object {
      "a": 1,
      "b": "b",
      "c": true,
    }
  }
>
  <form
    className="form form-horizontal"
  >
    <DummyComp
      a={1}
      b="b"
      c={true}
    >
      <span>
        test
      </span>
    </DummyComp>
  </form>
</FormPanel>
`;
